pkg/attribute/bag.go

```diff
@@ -41,6 +41,12 @@ type Bag interface {
 	// Bytes returns the named attribute if it exists.
 	Bytes(name string) ([]uint8, bool)
 
+	// Foreach allows a caller to process all of the (key, value) pairs this bag knows about. If the provided func
+	// returns false iteration is stopped.
+	//
+	// The view into the values is provided in this form so that the bag can keep iteration thread safe.
+	ForEach(handler func(key string, val interface{}) bool)
```

Comment:
The current approach will lead to deadlock if the handler happens to cause code to call some bag functions. Additionally, given that we're catching panics around adapter code, it's possible that a locked bag will be locked, the handler will cause a panic, we'll catch the panic, and the lock will not be unlocked.

Would be more reliable to implement instead a method that returns a flat [] of the bag's content. It's slower, requires allocation, but it can be made 100% safe.